File: typhoonapproach/typhoonapproach.go
package typhoonapproach

type Route []int

func NewRoute(rain ...int) Route {
	return rain
}

type Problem struct {
	Limit  int
	Routes []Route
}

func NewProblem(limit int, routes []Route) Problem {
	return Problem{limit, routes}
}

func Solve(problem Problem) (bool, []Route) {
	results := []Route{}
	for _, route := range problem.Routes {
		if canPass(problem.Limit, route) {
			results = append(results, route)
		}
	}
	if len(results) > 0 {
		return true, results
	} else {
		return false, nil
	}
}

func canPass(limit int, route Route) bool {
	for _, rain := range route {
		if limit <= rain {
			return false
		}
	}
	return true
}

File: src/typhoon_approach.rs
struct Route {
    rain: Vec<u32>,
}

impl Route {
    fn new(rain:Vec<u32>) -> Route {
        Route{rain}
    }
}

struct Problem {
    limit: u32,
    routes: Vec<Route>,
}

impl Problem {
    fn new(limit:u32, routes:Vec<Route>) -> Problem {
        Problem{limit, routes}
    }

    fn solve(&self) -> Option<Vec<usize>>{
        let mut routes = Vec::new();
        for (i, route) in self.routes.iter().enumerate(){
            if self.can_pass(route) {
                routes.push(i);
            }
        }
        if routes.is_empty() {
            None
        } else {
            Some(routes)
        }
    }

    fn can_pass(&self, route: &Route) -> bool {
        for rain in &route.rain {
            if self.limit <= *rain {
                return false;
            }
        }
        return true;
    }
}

#[cfg(test)]
mod tests {
    use super::Problem;
    use super::Route;

    #[test]
    fn test0(){
        let routes = vec![Route::new(vec![100,200,20]), Route::new(vec![100,20,20]), Route::new(vec![500,20,20])];
        let problem = Problem::new(200, routes);
        let want = Some(vec![1]);
        let got = problem.solve();
        assert_eq!(want, got);
    }
}